src/embed/connection.rs

```diff
@@ -0,0 +1,256 @@
+use super::dispatch::Request;
+use super::types::*;
+use crate::dag;
+use crate::db;
+use async_fn::AsyncFn2;
+use async_std::stream::StreamExt;
+use async_std::sync::{Receiver, RecvError, RwLock};
+use futures::stream::futures_unordered::FuturesUnordered;
+use log::warn;
+use nanoserde::{DeJson, SerJson};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU32, Ordering};
+
+lazy_static! {
+    static ref TRANSACTION_COUNTER: AtomicU32 = AtomicU32::new(1);
+}
+
+type TxnMap<'a> = RwLock<HashMap<u32, RwLock<db::Write<'a>>>>;
+
+fn deserialize<T: DeJson>(data: &str) -> Result<T, String> {
+    match DeJson::deserialize_json(data) {
+        Ok(v) => Ok(v),
+        Err(e) => Err(format!("InvalidJson({})", e)),
+    }
+}
+
+enum UnorderedResult {
+    Request(Result<Request, RecvError>),
+    Stop(),
+    None(),
+}
+
+async fn connection_future<'a, 'b>(
+    rx: &Receiver<Request>,
+    store: &'a dag::Store,
+    txns: &'b TxnMap<'a>,
+    request: Option<Request>,
+) -> UnorderedResult {
+    let req = match request {
+        None => return UnorderedResult::Request(rx.recv().await),
+        Some(v) => v,
+    };
+    match req.rpc.as_str() {
+        "has" => execute(do_has, txns, req).await,
+        "get" => execute(do_get, txns, req).await,
+        "put" => execute(do_put, txns, req).await,
+        "openTransaction" => do_open(store, txns, req).await,
+        "commitTransaction" => do_commit(txns, req).await,
+        "closeTransaction" => do_abort(txns, req).await,
+        "close" => {
+            req.response.send(Ok("".into())).await;
+            return UnorderedResult::Stop();
+        }
+        _ => {
+            req.response
+                .send(Err(format!("Unsupported rpc name {}", req.rpc)))
+                .await
+        }
+    };
+    UnorderedResult::None()
```

Comment:
This function returns an UnorderedResult in all cases. It returns `Request` if it pulled a new RPC off the Receiver, `Done` if it was asked to run an that RPC and the result is that the loop should stop, and `None` otherwise. The statement above this is terminated by a semicolon, and so isn't an implicit return.